core/src/main/java/org/pentaho/di/core/util/StringUtil.java

```diff
@@ -0,0 +1,717 @@
+  }
+
+  public static String initCap( String st ) {
+    if ( st == null || st.trim().length() == 0 ) {
+      return "";
+    }
+
+    if ( st.substring( 0, 1 ).equals( st.substring( 0, 1 ).toUpperCase() ) ) {
+      // Already initially capitalized.
+      return st;
+    } else {
+      // Capitalize first character
+      return st.substring( 0, 1 ).toUpperCase() + st.substring( 1 );
+    }
+  }
+
+  public static double str2num( String pattern, String decimal, String grouping, String currency, String value ) throws KettleValueException {
+    // 0 : pattern
+    // 1 : Decimal separator
+    // 2 : Grouping separator
+    // 3 : Currency symbol
+
+    NumberFormat nf = NumberFormat.getInstance();
+    DecimalFormat df = (DecimalFormat) nf;
+    DecimalFormatSymbols dfs = new DecimalFormatSymbols();
+
+    if ( !Utils.isEmpty( pattern ) ) {
+      df.applyPattern( pattern );
+    }
+    if ( !Utils.isEmpty( decimal ) ) {
+      dfs.setDecimalSeparator( decimal.charAt( 0 ) );
+    }
+    if ( !Utils.isEmpty( grouping ) ) {
+      dfs.setGroupingSeparator( grouping.charAt( 0 ) );
+    }
+    if ( !Utils.isEmpty( currency ) ) {
+      dfs.setCurrencySymbol( currency );
+    }
+    try {
+      df.setDecimalFormatSymbols( dfs );
+      return df.parse( value ).doubleValue();
+    } catch ( Exception e ) {
+      String message = "Couldn't convert string to number " + e.toString();
+      if ( !isEmpty( pattern ) ) {
+        message += " pattern=" + pattern;
+      }
+      if ( !isEmpty( decimal ) ) {
+        message += " decimal=" + decimal;
+      }
+      if ( !isEmpty( grouping ) ) {
+        message += " grouping=" + grouping.charAt( 0 );
+      }
+      if ( !isEmpty( currency ) ) {
+        message += " currency=" + currency;
+      }
+      throw new KettleValueException( message );
+    }
+  }
+
+  /**
+   * Check if the string supplied is empty. A String is empty when it is null or when the length is 0
+   *
+   * @param string
+   *          The string to check
+   * @return true if the string supplied is empty
+   */
+  public static final boolean isEmpty( String string ) {
+    return string == null || string.length() == 0;
+  }
+
+  /**
+   * Check if the StringBuilder supplied is empty. A StringBuilder is empty when it is null or when the length is 0
+   *
+   * @param string
+   *          The StringBuilder to check
+   * @return true if the StringBuilder supplied is empty
+   */
+  public static final boolean isEmpty( StringBuilder string ) {
+    return string == null || string.length() == 0;
+  }
+
+  public static Date str2dat( String arg0, String arg1, String val ) throws KettleValueException {
+    SimpleDateFormat df = new SimpleDateFormat();
+
+    DateFormatSymbols dfs = new DateFormatSymbols();
+    if ( arg1 != null ) {
+      dfs.setLocalPatternChars( arg1 );
+    }
+    if ( arg0 != null ) {
+      df.applyPattern( arg0 );
+    }
+
+    try {
+      return df.parse( val );
+    } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)